File: node/ts/test/chat/AuthMessagesServiceTest.ts
//
// Copyright 2026 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import { config, expect, use } from 'chai';
import chaiAsPromised from 'chai-as-promised';
import { Buffer } from 'node:buffer';

import * as Native from '../../Native.js';
import * as util from '../util.js';
import { AuthMessagesService, TokioAsyncContext } from '../../net.js';
import { connectAuth } from './ServiceTestUtils.js';
import { ErrorCode, LibSignalErrorBase } from '../../Errors.js';
import { FakeChatRemote, InternalRequest } from '../../net/FakeChat.js';
import { Aci } from '../../Address.js';
import { PlaintextContent } from '../../index.js';

use(chaiAsPromised);

util.initLogger();
config.truncateThreshold = 0;

describe('AuthMessagesService', () => {
  describe('getUploadForm', () => {
    it('works correctly', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectAuth<AuthMessagesService>(tokio);
      const responseFuture = chat.getUploadForm({ uploadSize: 42n });
      const request = await fakeRemote.assertReceiveIncomingRequest();
      expect(request.verb).to.eq('GET');
      expect(request.path).to.eq('/v4/attachments/form/upload?uploadLength=42');
      expect(request.headers.size).to.eq(0);
      expect(request.body.length).to.eq(0);
      fakeRemote.sendReplyTo(request, {
        status: 200,
        message: 'OK',
        headers: ['content-type: application/json'],
        body: new TextEncoder().encode(
          JSON.stringify({
            cdn: 123,
            key: 'abcde',
            headers: { one: 'val1', two: 'val2' },
            signedUploadLocation: 'http://example.org/upload',
          })
        ),
      });
      const response = await responseFuture;
      expect(response).to.deep.eq({
        cdn: 123,
        key: 'abcde',
        headers: new Map([
          ['one', 'val1'],
          ['two', 'val2'],
        ]),
        signedUploadUrl: new URL('http://example.org/upload'),
      });
    });
    it('throws on upload too large', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectAuth<AuthMessagesService>(tokio);
      const responseFuture = chat.getUploadForm({ uploadSize: 42n });
      const request = await fakeRemote.assertReceiveIncomingRequest();
      expect(request.verb).to.eq('GET');
      expect(request.path).to.eq('/v4/attachments/form/upload?uploadLength=42');
      expect(request.headers.size).to.eq(0);
      expect(request.body.length).to.eq(0);
      fakeRemote.sendReplyTo(request, {
        status: 413,
        message: 'Content Too Large',
      });
      await expect(responseFuture)
        .to.eventually.be.rejectedWith(LibSignalErrorBase)
        .and.deep.include({
          code: ErrorCode.UploadTooLarge,
        });
    });
  });

  describe('1:1 messages', () => {
    const recipientUuid = '4fcfe887-a600-40cd-9ab7-fd2a695e9981';

    async function sendTestMessage(
      chat: AuthMessagesService,
      syncMessage: boolean,
      fakeRemote: FakeChatRemote
    ): Promise<[Promise<void>, InternalRequest]> {
      const timestamp = 1700000000000;
      const contents = [
        {
          deviceId: 1,
          registrationId: 11,
          contents: PlaintextContent.deserialize(
            Uint8Array.of(0xc0, 1, 2, 3, 0x80)
          ).asCiphertextMessage(),
        },
        {
          deviceId: 2,
          registrationId: 22,
          contents: PlaintextContent.deserialize(
            Uint8Array.of(0xc0, 4, 5, 6, 0x80)
          ).asCiphertextMessage(),
        },
      ];
      const responseFuture = syncMessage
        ? chat.sendSyncMessage({ timestamp, contents, urgent: true })
        : chat.sendMessage({
            destination: Aci.fromUuid(recipientUuid),
            timestamp,
            contents,
            onlineOnly: false,
            urgent: true,
          });

      // Get the incoming request from the fake remote
      const request = await fakeRemote.assertReceiveIncomingRequest();

      expect(request.verb).to.eq('PUT');
      expect(request.path).to.eq(
        `/v1/messages/${
          syncMessage
            ? FakeChatRemote.FAKE_AUTH_CONNECT_SELF_UUID
            : recipientUuid
        }`
      );
      expect(request.headers).to.deep.eq(
        new Map([['content-type', 'application/json']])
      );
      expect(JSON.parse(new TextDecoder().decode(request.body))).to.deep.eq({
        messages: [
          {
            type: 8,
            destinationDeviceId: 1,
            destinationRegistrationId: 11,
            content: 'wAECA4A=',
          },
          {
            type: 8,
            destinationDeviceId: 2,
            destinationRegistrationId: 22,
            content: 'wAQFBoA=',
          },
        ],
        online: false,
        urgent: true,
        timestamp: 1700000000000,
      });

      return [responseFuture, request];
    }

    it('can send', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectAuth<AuthMessagesService>(tokio);

      for (const syncMessage of [false, true]) {
        const [responseFuture, request] = await sendTestMessage(
          chat,
          syncMessage,
          fakeRemote
        );

        fakeRemote.sendReplyTo(request, {
          status: 200,
          message: 'OK',
          headers: ['content-type: application/json'],
          body: Buffer.from('{}'),
        });

        await responseFuture;
      }
    });

    it('can handle NotFound', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectAuth<AuthMessagesService>(tokio);

      const [responseFuture, request] = await sendTestMessage(
        chat,
        false,
        fakeRemote
      );

      fakeRemote.sendReplyTo(request, {
        status: 404,
        message: 'Not Found',
      });

      await expect(responseFuture)
        .to.eventually.be.rejectedWith(LibSignalErrorBase)
        .and.deep.include({
          code: ErrorCode.ServiceIdNotFound,
        });
    });

    it('can handle a mismatched device error', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectAuth<AuthMessagesService>(tokio);

      const [responseFuture, request] = await sendTestMessage(
        chat,
        false,
        fakeRemote
      );

      fakeRemote.sendReplyTo(request, {
        status: 409,
        message: 'Conflict',
        headers: ['content-type: application/json'],
        body: Buffer.from(
          JSON.stringify({
            missingDevices: [4, 5],
            extraDevices: [40, 50],
          })
        ),
      });

      await expect(responseFuture)
        .to.eventually.be.rejectedWith(LibSignalErrorBase)
        .and.deep.include({
          code: ErrorCode.MismatchedDevices,
          entries: [
            {
              account: Aci.fromUuid(recipientUuid),
              missingDevices: [4, 5],
              extraDevices: [40, 50],
              staleDevices: [],
            },
          ],
        });
    });

    it('can handle a stale device error', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectAuth<AuthMessagesService>(tokio);

      const [responseFuture, request] = await sendTestMessage(
        chat,
        false,
        fakeRemote
      );

      fakeRemote.sendReplyTo(request, {
        status: 410,
        message: 'Gone',
        headers: ['content-type: application/json'],
        body: Buffer.from(
          JSON.stringify({
            staleDevices: [4, 5],
          })
        ),
      });

      await expect(responseFuture)
        .to.eventually.be.rejectedWith(LibSignalErrorBase)
        .and.deep.include({
          code: ErrorCode.MismatchedDevices,
          entries: [
            {
              account: Aci.fromUuid(recipientUuid),
              missingDevices: [],
              extraDevices: [],
              staleDevices: [4, 5],
            },
          ],
        });
    });

    it('can handle challenges', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectAuth<AuthMessagesService>(tokio);

      const [responseFuture, request] = await sendTestMessage(
        chat,
        false,
        fakeRemote
      );

      fakeRemote.sendReplyTo(request, {
        status: 428,
        message: 'Precondition Required',
        headers: ['content-type: application/json'],
        body: Buffer.from(
          JSON.stringify({
            token: 'zzz',
            options: ['captcha'],
          })
        ),
      });

      await expect(responseFuture)
        .to.eventually.be.rejectedWith(LibSignalErrorBase)
        .and.deep.include({
          code: ErrorCode.RateLimitChallengeError,
          token: 'zzz',
          options: new Set(['captcha']),
        });
    });
  });
});

File: java/client/src/test/java/org/signal/libsignal/net/AuthMessagesServiceTest.kt
//
// Copyright 2026 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.net

import kotlinx.serialization.json.Json
import org.junit.Test
import org.signal.libsignal.internal.TokioAsyncContext
import org.signal.libsignal.protocol.ServiceId.Aci
import org.signal.libsignal.protocol.message.PlaintextContent
import java.net.URI
import java.nio.charset.StandardCharsets
import java.util.UUID
import java.util.concurrent.Future
import java.util.concurrent.TimeUnit
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertIs

class AuthMessagesServiceTest {
  private val recipientUuid = UUID.fromString("4FCFE887-A600-40CD-9AB7-FD2A695E9981")

  @Test
  fun testGetUploadForm() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      AuthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
      )

    val service = AuthMessagesService(chat)
    val responseFuture = service.getUploadForm(42)
    val (request, requestId) = fakeRemote.getNextIncomingRequest().get(1, TimeUnit.SECONDS)
    assertEquals("GET", request.method)
    assertEquals("/v4/attachments/form/upload?uploadLength=42", request.pathAndQuery)
    assertEquals(0, request.headers.size)
    assertEquals(0, request.body.size)
    fakeRemote.sendResponse(
      requestId,
      200,
      "OK",
      arrayOf("content-type: application/json"),
      """
        {
          "cdn":123,
          "key":"abcde",
          "headers":{"one":"val1","two":"val2"},
          "signedUploadLocation":"http://example.org/upload"
        }
      """.encodeToByteArray(),
    )
    val result = responseFuture.get()
    val successResult = assertIs<RequestResult.Success<UploadForm>>(result)
    assertEquals(
      UploadForm(
        cdn = 123,
        key = "abcde",
        headers = mapOf("one" to "val1", "two" to "val2"),
        signedUploadUrl = URI("http://example.org/upload"),
      ),
      successResult.result,
    )
  }

  @Test
  fun testGetUploadFormTooLarge() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      AuthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
      )

    val service = AuthMessagesService(chat)
    val responseFuture = service.getUploadForm(42)
    val (request, requestId) = fakeRemote.getNextIncomingRequest().get(1, TimeUnit.SECONDS)
    assertEquals("GET", request.method)
    assertEquals("/v4/attachments/form/upload?uploadLength=42", request.pathAndQuery)
    assertEquals(0, request.headers.size)
    assertEquals(0, request.body.size)
    fakeRemote.sendResponse(
      requestId,
      413,
      "Content Too Large",
      arrayOf(),
      byteArrayOf(),
    )
    val error = assertIs<RequestResult.NonSuccess<UploadTooLargeException>>(responseFuture.get()).error
    assertIs<UploadTooLargeException>(error)
  }

  private fun sendTestMessage(
    chat: AuthenticatedChatConnection,
    syncMessage: Boolean,
    fakeRemote: FakeChatRemote,
  ): Pair<Future<out RequestResult<Unit, BadRequestError>>, Long> {
    val messagesService = AuthMessagesService(chat)

    val timestamp = 1700000000000L
    val expectedBody =
      Json.parseToJsonElement(
        """
        {
          "messages": [
            {
              "type": 8,
              "destinationDeviceId": 1,
              "destinationRegistrationId": 11,
              "content": "wAECA4A="
            },
            {
              "type": 8,
              "destinationDeviceId": 2,
              "destinationRegistrationId": 22,
              "content": "wAQFBoA="
            }
          ],
          "online": false,
          "urgent": true,
          "timestamp": 1700000000000
        }
        """,
      )

    val messages =
      listOf(
        SingleOutboundUnsealedMessage(1, 11, PlaintextContent(byteArrayOf(0xC0.toByte(), 1, 2, 3, 0x80.toByte()))),
        SingleOutboundUnsealedMessage(2, 22, PlaintextContent(byteArrayOf(0xC0.toByte(), 4, 5, 6, 0x80.toByte()))),
      )
    val responseFuture =
      if (syncMessage) {
        messagesService.sendSyncMessage(
          timestamp,
          messages,
          urgent = true,
        )
      } else {
        messagesService.sendMessage(
          Aci(recipientUuid),
          timestamp,
          messages,
          onlineOnly = false,
          urgent = true,
        )
      }

    // Get the incoming request from the fake remote
    val (request, requestId) = fakeRemote.getNextIncomingRequest().get()

    assertEquals("PUT", request.method)
    val expectedUuid = if (syncMessage) FakeChatRemote.FAKE_AUTH_CONNECT_SELF_UUID else recipientUuid
    assertEquals("/v1/messages/$expectedUuid", request.pathAndQuery)
    assertEquals(
      mapOf("content-type" to "application/json"),
      request.headers,
    )
    assertEquals(expectedBody, Json.parseToJsonElement(request.body.toString(Charsets.UTF_8)))

    return Pair(responseFuture, requestId)
  }

  @Test
  fun testSendMessageSuccess() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      AuthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        emptyArray(),
      )

    listOf(
      false,
      true,
    ).forEach { syncMessage ->
      val (responseFuture, requestId) = sendTestMessage(chat, syncMessage, fakeRemote)

      fakeRemote.sendResponse(
        requestId,
        200,
        "OK",
        arrayOf("content-type: application/json"),
        "{}".toByteArray(),
      )

      // Verify the result
      val result = responseFuture.get()
      assertIs<RequestResult.Success<Unit>>(result)
    }
  }

  @Test
  fun testSendMessageNotFound() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      AuthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        emptyArray(),
      )

    val (responseFuture, requestId) = sendTestMessage(chat, syncMessage = false, fakeRemote)
    fakeRemote.sendResponse(
      requestId,
      404,
      "Not Found",
      arrayOf(),
      byteArrayOf(),
    )

    // Verify the result
    val result = responseFuture.get()
    val nonSuccessResult = assertIs<RequestResult.NonSuccess<UnsealedSendFailure>>(result)
    assertIs<ServiceIdNotFoundException>(nonSuccessResult.error)
  }

  @Test
  fun testSendMessageMismatchedDevices() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      AuthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        emptyArray(),
      )

    val (responseFuture, requestId) = sendTestMessage(chat, syncMessage = false, fakeRemote)

    val jsonResponse =
      """
      {
        "missingDevices": [4, 5],
        "extraDevices": [40, 50]
      }
      """.trimIndent()

    fakeRemote.sendResponse(
      requestId,
      409,
      "Conflict",
      arrayOf("content-type: application/json"),
      jsonResponse.toByteArray(StandardCharsets.UTF_8),
    )

    // Verify the result
    val result = responseFuture.get()
    val nonSuccessResult = assertIs<RequestResult.NonSuccess<UnsealedSendFailure>>(result)
    val mismatchedDevices = assertIs<MismatchedDeviceException>(nonSuccessResult.error)
    assertContentEquals(
      arrayOf(
        MismatchedDeviceException.Entry(
          Aci(recipientUuid),
          missingDevices = intArrayOf(4, 5),
          extraDevices = intArrayOf(40, 50),
        ),
      ),
      mismatchedDevices.entries,
    )
  }

  @Test
  fun testSendMessageStaleDevices() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      AuthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        emptyArray(),
      )

    val (responseFuture, requestId) = sendTestMessage(chat, syncMessage = false, fakeRemote)

    val jsonResponse =
      """
      {
        "staleDevices": [4, 5]
      }
      """.trimIndent()

    fakeRemote.sendResponse(
      requestId,
      410,
      "Gone",
      arrayOf("content-type: application/json"),
      jsonResponse.toByteArray(StandardCharsets.UTF_8),
    )

    // Verify the result
    val result = responseFuture.get()
    val nonSuccessResult = assertIs<RequestResult.NonSuccess<UnsealedSendFailure>>(result)
    val mismatchedDevices = assertIs<MismatchedDeviceException>(nonSuccessResult.error)
    assertContentEquals(
      arrayOf(
        MismatchedDeviceException.Entry(
          Aci(recipientUuid),
          staleDevices = intArrayOf(4, 5),
        ),
      ),
      mismatchedDevices.entries,
    )
  }

  @Test
  fun testSendMessageCaptcha() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      AuthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        emptyArray(),
      )

    val (responseFuture, requestId) = sendTestMessage(chat, syncMessage = false, fakeRemote)

    val jsonResponse =
      """
      {
        "token": "zzz",
        "options": ["captcha"]
      }
      """.trimIndent()

    fakeRemote.sendResponse(
      requestId,
      428,
      "Precondition Required",
      arrayOf("content-type: application/json"),
      jsonResponse.toByteArray(StandardCharsets.UTF_8),
    )

    // Verify the result
    val result = responseFuture.get()
    val nonSuccessResult = assertIs<RequestResult.NonSuccess<UnsealedSendFailure>>(result)
    val challengeException = assertIs<RateLimitChallengeException>(nonSuccessResult.error)
    assertEquals("zzz", challengeException.token)
    assertEquals(setOf(ChallengeOption.CAPTCHA), challengeException.options)
  }
}
